datafusion/functions/src/datetime/to_local_time.rs

```diff
@@ -0,0 +1,601 @@
+                            transform_array::<TimestampMillisecondType>(&array, tz)
+                        }
+                        Timestamp(Second, Some(tz)) => {
+                            transform_array::<TimestampSecondType>(&array, tz)
+                        }
+                        _ => {
+                            exec_err!("to_local_time function requires timestamp argument in array, got {:?}", array.data_type())
+                        }
+                    }
+                }
+                _ => {
+                    exec_err!(
+                        "to_local_time function requires timestamp argument, got {:?}",
+                        time_value.data_type()
+                    )
+                }
+            },
+            _ => {
+                exec_err!(
+                    "to_local_time function requires timestamp argument, got {:?}",
+                    arg_type
+                )
+            }
+        }
+    }
+}
+
+/// This function converts a timestamp with a timezone to a timestamp without a timezone.
+/// The display value of the adjusted timestamp remain the same, but the underlying timestamp
+/// representation is adjusted according to the relative timezone offset to UTC.
+///
+/// This function uses chrono to handle daylight saving time changes.
+///
+/// For example,
+///
+/// ```text
+/// '2019-03-31T01:00:00Z'::timestamp at time zone 'Europe/Brussels'
+/// ```
+///
+/// is displayed as follows in datafusion-cli:
+///
+/// ```text
+/// 2019-03-31T01:00:00+01:00
+/// ```
+///
+/// and is represented in DataFusion as:
+///
+/// ```text
+/// TimestampNanosecond(Some(1_553_990_400_000_000_000), Some("Europe/Brussels"))
+/// ```
+///
+/// To strip off the timezone while keeping the display value the same, we need to
+/// adjust the underlying timestamp with the timezone offset value using `adjust_to_local_time()`
+///
+/// ```text
+/// adjust_to_local_time(1_553_990_400_000_000_000, "Europe/Brussels") --> 1_553_994_000_000_000_000
+/// ```
+///
+/// The difference between `1_553_990_400_000_000_000` and `1_553_994_000_000_000_000` is
+/// `3600_000_000_000` ns, which corresponds to 1 hour. This matches with the timezone
+/// offset for "Europe/Brussels". Consequently, DataFusion can represent the timestamp in
+/// local time (with no offset or timezone information) as
+///
+/// ```text
+/// TimestampNanosecond(Some(1_553_994_000_000_000_000), None)
+/// ```
+///
+/// which is displayed as follows in datafusion-cli:
+///
+/// ```text
+/// 2019-03-31T01:00:00
+/// ```
+fn adjust_to_local_time<T: ArrowTimestampType>(ts: i64, timezone: &str) -> i64 {
+    let date_time = match T::UNIT {
+        Nanosecond => Utc.timestamp_nanos(ts),
+        Microsecond => Utc.timestamp_micros(ts).unwrap(), // TODO chunchun: replace unwrap
+        Millisecond => Utc.timestamp_millis_opt(ts).unwrap(),
+        Second => Utc.timestamp_opt(ts, 0).unwrap(),
+    };
+
+    let tz: Tz = timezone.parse().unwrap();
```

Comment:
1. I think the timezone parsing should probably also return an error rather than unwrap

Reply:
2. As written this is going to parse the timezone for each row, which will likely be much slower than needed

I think you could do this timezone parsing once in `transform_array` and then pass the parsed timezone into `adjust_to_local_time`

Reply:
Thank you for the suggestion! I updated them in two commits: 

- https://github.com/apache/datafusion/pull/11347/commits/043c9971a4ab9106b5d3002ad8ee6b84a76967f0 -- parse timezone once
- https://github.com/apache/datafusion/pull/11347/commits/afe619f057eb4c7830f1fa01933527d21fa30d61 -- handle the error rather than unwrap